File: 12 - Chain of Responsibility/Python/method_chain.py
class Creature:
    def __init__(self, name, attack, defense) -> None:
        self.name = name
        self.attack = attack
        self.defense = defense
    
    def __str__(self) -> str:
        return f'{self.name} ({self.attack}/{self.defense})'


class CreatureModifier:
    def __init__(self, creature) -> None:
        self.creature = creature
        self.next_modifier = None
    
    def add_modifier(self, modifier):
        if self.next_modifier:
            self.next_modifier.add_modifier(modifier)
        else:
            self.next_modifier = modifier

    def handle(self):
        if self.next_modifier:
            self.next_modifier.handle()


class DoubleAttackModifier(CreatureModifier):
    def handle(self):
        print(f'Doubling {self.creature.name}\'s attack.')
        self.creature.attack *= 2
        super().handle()

class IncreaseDefenseModifier(CreatureModifier):
    def handle(self):
        print(f'Increase {self.creature.name}\'s defense.')
        self.creature.defense *= 4
        super().handle()

class NoBonusModifier(CreatureModifier):
    def handle(self):
        print("No bonuses for you")


if __name__ == "__main__":
    goblin = Creature('Goblin', 1, 1)
    print(goblin)

    root = CreatureModifier(goblin)
    root.add_modifier(DoubleAttackModifier(goblin))
    root.add_modifier(DoubleAttackModifier(goblin))
    root.add_modifier(NoBonusModifier(goblin))
    root.add_modifier(IncreaseDefenseModifier(goblin))
    root.handle()
    print(goblin)

File: 12 - Chain of Responsibility/Dart/method_chain.dart
void main() {
  Creature goblin = Creature('Goblin', 1, 1);
  print(goblin);

  CreatureModifier root = CreatureModifier(goblin);
  root.addModofier(DoubleAttackModifier(goblin));
  root.addModofier(DoubleAttackModifier(goblin));
  root.addModofier(NoBonusModifier(goblin));
  root.addModofier(IncreaseDefenseModifier(goblin));
  root.handle();
  print(goblin);
}

class Creature {
  String name;
  int attack;
  int defense;

  Creature(this.name, this.attack, this.defense);

  @override
  String toString() => "$name ($attack/$defense)";
}

class CreatureModifier {
  Creature creature;
  CreatureModifier? nextModifier;

  CreatureModifier(this.creature);

  void handle() {
    if (nextModifier != null) {
      nextModifier!.handle();
    }
  }

  void addModofier(CreatureModifier modifier) {
    if (nextModifier != null) {
      nextModifier!.addModofier(modifier);
    } else {
      nextModifier = modifier;
    }
  }
}

class DoubleAttackModifier extends CreatureModifier {
  Creature c;
  DoubleAttackModifier(this.c) : super(c);

  @override
  void handle() {
    print('Doubling ${creature.name}\'s attack.');
    creature.attack *= 2;
    super.handle();
  }
}

class IncreaseDefenseModifier extends CreatureModifier {
  Creature c;
  IncreaseDefenseModifier(this.c) : super(c);

  @override
  void handle() {
    print('Increase ${creature.name}\'s defense.');
    creature.defense *= 4;
    super.handle();
  }
}

class NoBonusModifier extends CreatureModifier {
  Creature c;
  NoBonusModifier(this.c) : super(c);

  @override
  void handle() {
    print('No bonuses for you');
  }
}
